src/MongoDB.Driver.Core/Core/Configuration/ClusterBuilder.cs

```diff
@@ -248,5 +210,85 @@ public ClusterBuilder Subscribe(IEventSubscriber subscriber)
             _eventAggregator.Subscribe(subscriber);
             return this;
         }
+
+        // private methods
+        private IClusterFactory CreateClusterFactory()
+        {
+            var serverFactory = CreateServerFactory();
+
+            return new ClusterFactory(
+                _clusterSettings,
+                serverFactory,
+                _eventAggregator);
+        }
+
+        private IConnectionPoolFactory CreateConnectionPoolFactory()
+        {
+            var streamFactory = CreateTcpStreamFactory(_tcpStreamSettings);
+
+            var connectionFactory = new BinaryConnectionFactory(
+                _connectionSettings,
+                streamFactory,
+                _eventAggregator);
+
+            return new ExclusiveConnectionPoolFactory(
+                _connectionPoolSettings,
+                connectionFactory,
+                _eventAggregator);
+        }
+
+        private ServerFactory CreateServerFactory()
+        {
+            var connectionPoolFactory = CreateConnectionPoolFactory();
+            var serverMonitorFactory = CreateServerMonitorFactory();
+
+            return new ServerFactory(
+                _clusterSettings.ConnectionMode,
+                _serverSettings,
+                connectionPoolFactory,
+                serverMonitorFactory,
+                _eventAggregator);
+        }
+
+        private IServerMonitorFactory CreateServerMonitorFactory()
+        {
+            var serverMonitorConnectionSettings = _connectionSettings
+                .With(authenticators: new IAuthenticator[] { });
+
+            var serverMonitorConnectTimeout = _tcpStreamSettings.ConnectTimeout;
+            if (serverMonitorConnectTimeout == TimeSpan.Zero || serverMonitorConnectTimeout == Timeout.InfiniteTimeSpan)
+            {
+                serverMonitorConnectTimeout = TimeSpan.FromSeconds(30);
+            }
+            var serverMonitorTcpStreamSettings = new TcpStreamSettings(_tcpStreamSettings)
+                .With(
+                    connectTimeout: serverMonitorConnectTimeout,
+                    readTimeout: _serverSettings.HeartbeatTimeout,
```

Comment:
Sorry, just thought of one more thing.  Can we disallow zero or infinite heartbeat timeouts as well?  Are we already?